test/core/end2end/cq_verifier.h

```diff
@@ -23,44 +23,68 @@
-
-typedef struct cq_verifier cq_verifier;
-
-/* construct/destroy a cq_verifier */
-cq_verifier* cq_verifier_create(grpc_completion_queue* cq);
-void cq_verifier_destroy(cq_verifier* v);
-
-/* ensure all expected events (and only those events) are present on the
-   bound completion queue within \a timeout_sec */
-void cq_verify(cq_verifier* v, int timeout_sec = 10);
-
-/* ensure that the completion queue is empty */
-void cq_verify_empty(cq_verifier* v);
-
-/* ensure that the completion queue is empty, waiting up to \a timeout secs. */
-void cq_verify_empty_timeout(cq_verifier* v, int timeout_sec);
-
-/* Various expectation matchers
-   Any functions taking ... expect a NULL terminated list of key/value pairs
-   (each pair using two parameter slots) of metadata that MUST be present in
-   the event. */
-void cq_expect_completion(cq_verifier* v, const char* file, int line, void* tag,
-                          bool success);
-/* If the \a tag is seen, \a seen is set to true. */
-void cq_maybe_expect_completion(cq_verifier* v, const char* file, int line,
-                                void* tag, bool success, bool* seen);
-void cq_expect_completion_any_status(cq_verifier* v, const char* file, int line,
-                                     void* tag);
-#define CQ_EXPECT_COMPLETION(v, tag, success) \
-  cq_expect_completion(v, __FILE__, __LINE__, tag, success)
-#define CQ_MAYBE_EXPECT_COMPLETION(v, tag, success, seen) \
-  cq_maybe_expect_completion(v, __FILE__, __LINE__, tag, success, seen)
-#define CQ_EXPECT_COMPLETION_ANY_STATUS(v, tag) \
-  cq_expect_completion_any_status(v, __FILE__, __LINE__, tag)
+namespace grpc_core {
+
+// A CqVerifier can verify that expected events arrive in a timely fashion
+// on a single completion queue
+class CqVerifier {
+ public:
+  explicit CqVerifier(grpc_completion_queue* cq);
+  ~CqVerifier();
+
+  CqVerifier(const CqVerifier&) = delete;
+  CqVerifier& operator=(const CqVerifier&) = delete;
+
+  // Ensure all expected events (and only those events) are present on the
+  // bound completion queue within \a timeout.
+  void Verify(Duration timeout = Duration::Seconds(10));
+
+  // Ensure that the completion queue is empty, waiting up to \a timeout.
+  void VerifyEmpty(Duration timeout = Duration::Seconds(1));
+
+  // ExpectedResult - if the tag is received, set *seen to true (if seen is
+  // non-null).
+  struct Maybe {
+    bool* seen = nullptr;
+  };
+  // ExpectedResult - expect the tag, but its result may be true or false.
+  // Store the result in result (if result is non-null).
+  struct AnyStatus {
+    bool* result = nullptr;
+  };
+
+  using ExpectedResult = absl::variant<bool, Maybe, AnyStatus>;
+
+  /* Various expectation matchers
+     Any functions taking ... expect a NULL terminated list of key/value pairs
+     (each pair using two parameter slots) of metadata that MUST be present in
+     the event. */
+  void Expect(DebugLocation location, void* tag, ExpectedResult result);
```

Comment:
suggest a future iteration

Reply:
I'll leave it up to you, but I will point out that this is going to be a regression in behavior caused by this PR.  The current code shows file and line numbers for failures even in opt builds.  I think without that information, it may be fairly difficult to debug test failures.